vm_method.c

```diff
@@ -303,27 +303,27 @@ rb_clear_method_cache(VALUE klass_or_module, ID mid)
 }
 
 static int
-invalidate_all_refinement_cc(void *vstart, void *vend, size_t stride, void *data)
-{
-    VALUE v = (VALUE)vstart;
-    for (; v != (VALUE)vend; v += stride) {
-        void *ptr = rb_asan_poisoned_object_p(v);
-        rb_asan_unpoison_object(v, false);
-
-        if (RBASIC(v)->flags) { // liveness check
-            if (imemo_type_p(v, imemo_callcache)) {
-                const struct rb_callcache *cc = (const struct rb_callcache *)v;
-                if (vm_cc_refinement_p(cc) && cc->klass) {
-                    vm_cc_invalidate(cc);
-                }
-            }
-        }
+invalidate_cc_refinement(st_data_t key, st_data_t value, st_data_t data)
+{
+    VALUE v = (VALUE)value;
+    void *ptr = rb_asan_poisoned_object_p(v);
+    rb_asan_unpoison_object(v, false);
+
+    if (RBASIC(v)->flags) { // liveness check
```

Comment:
You need `rb_gc_pointer_to_heap_p(v) && !rb_objspace_garbage_object_p(v)` before you can dereference `v` now, since it's in a weak table. By the time this runs `v` could point to an unmapped memory. cc @peterzhu2118

Reply:
Thank you for the review! I've applied the changes and force-pushed the update.

Reply:
Can you rerun the benchmarks? `rb_gc_pointer_to_heap_p` is a pretty slow API, and will be slower the larger the heap (e.g. in a system with a lot of live objects such as a Rails app).

Reply:
Updated the benchmark in the description 👍 
<img width="630" alt="スクリーンショット 2025-04-09 0 07 31" src="https://github.com/user-attachments/assets/3fac43d2-7572-47c9-97dd-a3d9ed462690" />

It's an unofficial benchmark, but the performance improvement is clear enough in our Rails application where the original performance regression was discovered and I'm happy with the result.

```
> $ bundle exec rails r 'puts Benchmark.measure { Rails.application.eager_load! }'
> before 12.689740   1.162582  13.852322 ( 14.480201)
> after   2.317027   0.605092   2.922119 (  3.348900)

> $ touch app/controllers/application_controller.rb && time curl http://localhost:3000/
> before 0.00s user 0.00s system 0% cpu 11.263 total
> after  0.00s user 0.01s system 0% cpu 2.503 total
```